File: 2022/day2-rockPaperScissors.js
const { readFile } = require('../utils');

(async () => {
  const input = await readFile('day2.txt');

  // A, X: Rock
  // B, Y: Paper
  // C, Z: Scissors
  const roundScoreMap = {
    'A X': 1 + 3,
    'A Y': 2 + 6,
    'A Z': 3 + 0,
    'B X': 1 + 0,
    'B Y': 2 + 3,
    'B Z': 3 + 6,
    'C X': 1 + 6,
    'C Y': 2 + 0,
    'C Z': 3 + 3,
  };

  // A: Rock
  // B: Paper
  // C: Scissors
  // X: Lose
  // Y: Draw
  // Z: Win
  const roundScoreMapPart2 = {
    'A X': 3 + 0,
    'A Y': 1 + 3,
    'A Z': 2 + 6,
    'B X': 1 + 0,
    'B Y': 2 + 3,
    'B Z': 3 + 6,
    'C X': 2 + 0,
    'C Y': 3 + 3,
    'C Z': 1 + 6,
  };

  console.log(rockPaperScissors(input, roundScoreMap));
  console.log(rockPaperScissors(input, roundScoreMapPart2));
})();

const rockPaperScissors = (input, roundScoreMap) => {
  return input.reduce((result, round) => {
    const score = roundScoreMap[round];
    return result + score;
  }, 0);
};

File: 2022/day2-rockPaperScissors.kt
import java.io.File

fun readInput(file: String, callback: ((line: String) -> Any)?): ArrayList<Any> {
  val input = ArrayList<Any>()
  File(file).reader().forEachLine { line ->
    val newLine = callback?.let { callback(line) } ?: line
    input.add(newLine)
  }
  return input
}

fun main() {
  val input = readInput("inputs/day2.txt", null).toArray(arrayOf<String>())
  val roundScoreMap =
      mapOf(
          "A X" to 1 + 3,
          "A Y" to 2 + 6,
          "A Z" to 3 + 0,
          "B X" to 1 + 0,
          "B Y" to 2 + 3,
          "B Z" to 3 + 6,
          "C X" to 1 + 6,
          "C Y" to 2 + 0,
          "C Z" to 3 + 3
      )
  val roundScoreMapPart2 =
      mapOf(
          "A X" to 3 + 0,
          "A Y" to 1 + 3,
          "A Z" to 2 + 6,
          "B X" to 1 + 0,
          "B Y" to 2 + 3,
          "B Z" to 3 + 6,
          "C X" to 2 + 0,
          "C Y" to 3 + 3,
          "C Z" to 1 + 6
      )
  println(rockPaperScissors(input, roundScoreMap))
  println(rockPaperScissors(input, roundScoreMapPart2))
}

fun rockPaperScissors(input: Array<String>, roundScoreMap: Map<String, Int>): Int {
  return input.fold(0) { result, round ->
    val score = roundScoreMap.get(round) ?: 0
    return@fold result + score
  }
}
